fdbserver/RestoreMaster.actor.h

```diff
@@ -65,50 +69,113 @@ struct VersionBatch {
+				TraceEvent(SevError, "FastRestore")
+				    .detail("SanityCheckApplierKeyRange", applierToRange.size())
+				    .detail("ApplierID", applier.second)
+				    .detail("Key1", applierToRange[applier.second])
+				    .detail("Key2", applier.first);
+				ret = false;
+			}
+		}
+		return ret;
+	}
+
+	void logApplierKeyRange() {
+		TraceEvent("FastRestore").detail("ApplierKeyRangeNum", rangeToApplier.size());
+		for (auto& applier : rangeToApplier) {
+			TraceEvent("FastRestore").detail("KeyRangeLowerBound", applier.first).detail("Applier", applier.second);
+		}
+	}
+};
+
+enum class RestoreAssetStatus { Loading, Loaded };
+
+enum class RestoreSendStatus { SendingLogs, SendedLogs, SendingRanges, SendedRanges };
+
+enum class RestoreApplyStatus { Applying, Applied };
 
-	int batchIndex;
+// Track restore progress of each RestoreAsset (RA) and
+// Use status to sanity check restore property, e.g., each RA should be processed exactly once.
+struct MasterBatchStatus : public ReferenceCounted<MasterBatchStatus> {
+	std::map<RestoreAsset, RestoreAssetStatus> raStatus;
+	std::map<UID, RestoreSendStatus> loadStatus;
+	std::map<UID, RestoreApplyStatus> applyStatus;
+
+	void addref() { return ReferenceCounted<MasterBatchStatus>::addref(); }
+	void delref() { return ReferenceCounted<MasterBatchStatus>::delref(); }
+
+	MasterBatchStatus() = default;
+	~MasterBatchStatus() = default;
+};
+
+struct RestoreMasterData : RestoreRoleData, public ReferenceCounted<RestoreMasterData> {
+	std::map<Version, VersionBatch> versionBatches; // key is the beginVersion of the version batch
 
 	Reference<IBackupContainer> bc; // Backup container is used to read backup files
 	Key bcUrl; // The url used to get the bc
 
-	IndexedSet<Key, int64_t> samples; // sample of range and log files
-	double samplesSize; // sum of the metric of all samples
+	std::map<int, Reference<MasterBatchData>> batch;
+	std::map<int, Reference<MasterBatchStatus>> batchStatus;
+	NotifiedVersion finishedBatch; // The highest batch index all appliers have applied mutations
+
+	AsyncVar<int> runningVersionBatches; // Currently running version batches
 
 	void addref() { return ReferenceCounted<RestoreMasterData>::addref(); }
 	void delref() { return ReferenceCounted<RestoreMasterData>::delref(); }
 
 	RestoreMasterData() {
 		role = RestoreRole::Master;
 		nodeID = UID();
-		batchIndex = 1; // starts with 1 because batchId (NotifiedVersion) in loaders and appliers start with 0
+		runningVersionBatches.set(0);
 	}
 
 	~RestoreMasterData() = default;
 
-	void resetPerVersionBatch() {
-		TraceEvent("FastRestore")
-		    .detail("RestoreMaster", "ResetPerVersionBatch")
-		    .detail("VersionBatchIndex", batchIndex);
-		samplesSize = 0;
-		samples.clear();
+	void initVersionBatch(int batchIndex) {
+		TraceEvent("FastRestoreMasterInitVersionBatch", id()).detail("VersionBatchIndex", batchIndex);
+	}
+
+	// Reset master data at the beginning of each restore request
+	void resetPerRestoreRequest() {
```

Comment:
Does `runningVersionBatches` need to be cleared in this function?

Reply:
Nope. The variable indicates the current outstanding version batches.
When the next restore request happens, it also decreases to 0.